test/blackbox-tests/test-cases/modules-expansion.t

```diff
@@ -0,0 +1,177 @@
+  File "dune", line 1, characters 18-32:
+  1 | (executable (name does_not_exist) (modules (:include gen/lst)))
+                        ^^^^^^^^^^^^^^
+  Error: The name "Does_not_exist" is not listed in the (modules) field of this
+  stanza.
+  [1]
+
+  $ cat >dune <<EOF
+  > (executable (name mod) (modules does_not_exist (:include gen/lst)))
+  > EOF
+
+  $ dune exec ./mod.exe
+  File "dune", line 1, characters 32-46:
+  1 | (executable (name mod) (modules does_not_exist (:include gen/lst)))
+                                      ^^^^^^^^^^^^^^
+  Error: Module Does_not_exist doesn't exist.
+  [1]
+
+  $ mv mod.ml mod2.ml
+  $ cat >dune <<EOF
+  > (executable (name mod) (modules (:include gen/lst)))
+  > EOF
+
+Locations are accurate even for generated files:
+
+  $ dune exec ./mod.exe
+  File "_build/default/gen/lst", line 1, characters 0-3:
+  1 | mod
+      ^^^
+  Error: Module Mod doesn't exist.
+  [1]
+
+  $ mv mod2.ml mod.ml
+
+Let's do some examples using libraries:
+
+  $ cat >dune <<EOF
+  > (library
+  >  (name lib)
+  >  (modules (:include gen/lst)))
+  > EOF
+
+  $ dune build --display short
+        ocamlc .lib.objs/byte/lib.{cmi,cmo,cmt}
+      ocamldep .lib.objs/lib__Mod.impl.d
+      ocamlopt .lib.objs/native/lib.{cmx,o}
+        ocamlc .lib.objs/byte/lib__Mod.{cmi,cmo,cmt}
+      ocamlopt .lib.objs/native/lib__Mod.{cmx,o}
+        ocamlc lib.cma
+      ocamlopt lib.{a,cmxa}
+      ocamlopt lib.cmxs
+
+We can also use special forms such as `%{read-lines:}`:
+
+  $ cat >dune <<EOF
+  > (library
+  >  (name lib)
+  >  (modules %{read-lines:gen/lst}))
+  > EOF
+
+  $ cat >gen/dune <<EOF
+  > (rule (with-stdout-to lst (echo "mod\nmod2\n")))
+  > EOF
+
+  $ touch mod2.ml
+
+  $ dune build --display short
+        ocamlc .lib.objs/byte/lib.{cmi,cmo,cmt}
+      ocamldep .lib.objs/lib__Mod2.impl.d
+      ocamlopt .lib.objs/native/lib.{cmx,o}
+        ocamlc .lib.objs/byte/lib__Mod.{cmi,cmo,cmt}
+        ocamlc .lib.objs/byte/lib__Mod2.{cmi,cmo,cmt}
+      ocamlopt .lib.objs/native/lib__Mod.{cmx,o}
+      ocamlopt .lib.objs/native/lib__Mod2.{cmx,o}
+        ocamlc lib.cma
+      ocamlopt lib.{a,cmxa}
+      ocamlopt lib.cmxs
+
+Next, we illustrate the issue mentioned above: the build dependencies must not
+live in the same directory as the containing stanza, otherwise a cycle
+appears. We need to handle this cycle gracefully and report it to the user.
+
+  $ cat >dune <<EOF
+  > (rule (with-stdout-to lst (echo "mod")))
+  > (executable (name mod) (modules (:include lst)))
+  > EOF
+
+  $ dune exec ./mod.exe
+  Internal error, please report upstream including the contents of _build/log.
+  Description:
+    ("internal dependency cycle",
+    { frames =
+        [ ("build-file", In_build_dir "default/lst")
+        ; ("<unnamed>", ())
+        ; ("<unnamed>", ())
+        ; ("load-dir", In_build_dir "default")
+        ]
+    })
+  Raised at Memo.Exec.exec_dep_node.(fun) in file "src/memo/memo.ml", line
```

Comment:
We don't include stack traces in the test output (they make the tests fail due to spurious reasons). See how we filter such stacktraces elsewhere with `awk`.

Reply:
The stack trace was meant to document the issue to aid reviewing, but I hope we will be able to get rid of it before the PR is merged :)

Reply:
I pushed a commit with an attempt for a more user-friendly stack trace ("Human readable stack trace in case of cycle").